### ISSUE TITLE
Multi-GPU

### ISSUE THREAD

**Issue by viperyl** [BOINC/boinc#5438]
**Describe the bug**
boinccmd show the tasks running on the different device, but all of task running the same gpu by nvidia-smi command

**Steps To Reproduce**

**Expected behavior**
nvidia device 1 task should to use device:1, rather than device:0

**Screenshots**
boinccmd --get_task_summary outputs
![Screenshot 2023-11-27 at 14 27 08](https://github.com/BOINC/boinc/assets/45485217/a486e09d-714c-4b7c-a8b1-d0f2506efb12)

nvidia-top outputs (same as nvidia-smi)
![Screenshot 2023-11-27 at 14 27 54](https://github.com/BOINC/boinc/assets/45485217/496f4f42-6064-41a6-ba65-21efc198e630)

cc_config.xml outputs
![Screenshot 2023-11-27 at 14 28 31](https://github.com/BOINC/boinc/assets/45485217/3a795e10-eed4-466b-9eec-033f6231b87a)

htop outpus (basic system infomation)
![Screenshot 2023-11-27 at 14 30 26](https://github.com/BOINC/boinc/assets/45485217/3236fc69-28b9-488c-8371-2cab6bd78007)

**System Information**
 - OS: Manajaro
 - BOINC Version: 7.24.1 x86_64-pc-linux-gnu

**Additional context**

**Comment by RichardHaselgrove** [comment]
This is likely to be a problem with the project science application (Numberfields@home).

A similar problem was recently introduced, and later fixed, by the Einstein@home project.

**Comment by RichardHaselgrove** [comment]
Negative confirmation, I'm afraid.

First screenshot shows a different project, with one task running on each GPU, as expected:

![image](https://github.com/BOINC/boinc/assets/14886436/0c8f6a8a-ad89-4407-8f3c-cfaa50e3b582)

Second screenshot shows Numberfields@home - also working properly:

![image](https://github.com/BOINC/boinc/assets/14886436/88e937de-8a21-4fc2-8905-51df284ec60f)

Final screenshot shows Numberfields@home device allocation to device 1 as expected.

![image](https://github.com/BOINC/boinc/assets/14886436/8c7cf6b0-4347-4c72-993b-dc01140d50d3)

Images taken from Linux Mint 20.3, running BOINC client 7.20.5 (last from PPA)

**Comment by viperyl** [comment]
> Negative confirmation, I'm afraid.
> 
> First screenshot shows a different project, with one task running on each GPU, as expected:
> 
> ![image](https://private-user-images.githubusercontent.com/14886436/285869949-0c8f6a8a-ad89-4407-8f3c-cfaa50e3b582.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwOTE0NDUsIm5iZiI6MTcwMTA5MTE0NSwicGF0aCI6Ii8xNDg4NjQzNi8yODU4Njk5NDktMGM4ZjZhOGEtYWQ4OS00NDA3LThmM2MtY2ZhYTUwZTNiNTgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI3VDEzMTkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlZTJkNmM1YjkwMzI3NzFkNDg3YzdiZDE3ZGZkMjBlYTMyM2IwNjMwODA4ZjIxNmJjMjRkZTY3NDU1NTUxNjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4wyiTY4pjQFJG1MDEjJ-BF_H00Dyky3eejcKrJ5Thi4)
> 
> Second screenshot shows Numberfields@home - also working properly:
> 
> ![image](https://private-user-images.githubusercontent.com/14886436/285870409-88e937de-8a21-4fc2-8905-51df284ec60f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwOTE0NDUsIm5iZiI6MTcwMTA5MTE0NSwicGF0aCI6Ii8xNDg4NjQzNi8yODU4NzA0MDktODhlOTM3ZGUtOGEyMS00ZmMyLTg5MDUtNTFkZjI4NGVjNjBmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI3VDEzMTkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzNDEwODVjMGRjMzE2MWQ0NTc2OTc2MjQ4ODhlMjM5ZmM3ZmIwNWNjZjU3YzJjNjAzNTEyYTgwY2FlN2YyYjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xRx8i-bRNuviL2i18bDCctns3qf4rMVRE7iUDl2RD5Q)
> 
> Final screenshot shows Numberfields@home device allocation to device 1 as expected.
> 
> ![image](https://private-user-images.githubusercontent.com/14886436/285871028-8c7cf6b0-4347-4c72-993b-dc01140d50d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwOTE0NDUsIm5iZiI6MTcwMTA5MTE0NSwicGF0aCI6Ii8xNDg4NjQzNi8yODU4NzEwMjgtOGM3Y2Y2YjAtNDM0Ny00YzcyLTk5M2ItZGMwMTE0MGQ1MGQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI3VDEzMTkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdhMzRiNDY0NDg3Y2Y4MDA3Y2I1NzVhNjM2MDIyOWU2N2JlZGE1NmNjMTQxMWQ0NmZjMjhhYzcyYWY4Y2I2NGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Ccs3ar5KJ62L_6X5JYev5n_OBSxOYXqEljMFx2f_flw)
> 
> Images taken from Linux Mint 20.3, running BOINC client 7.20.5 (last from PPA)

But on my server, all of CUDA task running on the device:0, rather than `boinccmd --get_task_summary` showed averaged running on seperate GPU. I will try to join Einstein@home for further testing.

**Comment by RichardHaselgrove** [comment]
Reference to the Einstein project administrator's acknowledgement that their similar problem had been caused by their science application:

https://einsteinathome.org/content/multi-gpu-not-consistantly-applied-boincs-request#comment-219034

Read before and after in the same thread for detection and resolution.

**Comment by viperyl** [comment]
> Reference to the Einstein project administrator's acknowledgement that their similar problem had been caused by their science application:
> 
> https://einsteinathome.org/content/multi-gpu-not-consistantly-applied-boincs-request#comment-219034
> 
> Read before and after in the same thread for detection and resolution.

thanks for you link, i have detailed read it, and check all of output log of bonic, every GPUs are correctly idenfied by bonic with differetn PCI id, everything looks right, but all of cuda computaton task still running on the first device (device:0).
Really strange, because I sucessfully t allocate array to differetn devices by python. I will try to downgrade bonic version later. (Should I use HDMI cable and GUI version bonic?)

**Comment by viperyl** [comment]
> Reference to the Einstein project administrator's acknowledgement that their similar problem had been caused by their science application:
> 
> https://einsteinathome.org/content/multi-gpu-not-consistantly-applied-boincs-request#comment-219034
> 
> Read before and after in the same thread for detection and resolution.

I skipped this problem by through Docker. At inital, I set all of GPU visiable for docker, but i met the same problem at the start, all of tasks running on the deivce:0. So I manually create four dockers, and assign different GPU for each docker, also euqally assign CPU for each docker. Now everything works great.

`docker ps -a` outputs
![Screenshot 2023-11-29 at 14 01 58](https://github.com/BOINC/boinc/assets/45485217/33cd681f-7e9b-42d2-907f-96216e256f39)

`nvitop` outputs
![Screenshot 2023-11-29 at 14 02 31](https://github.com/BOINC/boinc/assets/45485217/8278837f-0284-4e5f-9b7e-78f783598a14)

I have another no longer used quad 1080ti and quad 2080ti machines, should I use k8s to manage these machine? :D
